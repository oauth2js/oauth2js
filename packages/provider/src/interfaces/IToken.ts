import { Maybe } from '../helpers/types'
import { IClient } from './IClient'
import { IScope } from './IScope'
import { IUser } from './IUser'

export interface IToken extends OAuth2JS.Server.Token {
  accessToken: string
  accessTokenExpiresAt: Date
  refreshToken: Maybe<string>
  refreshTokenExpiresAt: Maybe<Date>
  scopes: IScope[]
  client: IClient
  user: IUser
}

declare global {
  namespace OAuth2JS {
    namespace Server {
      interface Token {}
    }
  }
}
