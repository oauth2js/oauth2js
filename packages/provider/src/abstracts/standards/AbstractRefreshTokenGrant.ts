import { Mixed } from '@prunus/mixin'

import { InvalidClient } from '../../errors/InvalidClient'
import { InvalidGrant } from '../../errors/InvalidGrant'
import {
  DEFAULT_ACCESS_TOKEN_LIFETIME,
  DEFAULT_REFRESH_TOKEN_LIFETIME,
} from '../../helpers/defaults'
import { Maybe, PromiseToo } from '../../helpers/types'
import { RequestLocation } from '../../interfaces/IRequest'
import { IToken } from '../../interfaces/IToken'
import { Context } from '../../models/Context'
import { UseClientId } from '../../traits/UseClientId'
import { UseTokenByRefreshToken } from '../../traits/UseTokenByRefreshToken'
import { AbstractGrant } from '../AbstractGrant'

export abstract class AbstractRefreshTokenGrant extends Mixed(
  AbstractGrant,
  UseTokenByRefreshToken,
  UseClientId,
) {
  protected accessTokenLifetime: number = DEFAULT_ACCESS_TOKEN_LIFETIME
  protected refreshTokenLifetime: number = DEFAULT_REFRESH_TOKEN_LIFETIME
  protected readonly clientIdLocation: RequestLocation = RequestLocation.Body

  protected readonly refreshTokenLocation: RequestLocation =
    RequestLocation.Body

  protected async createToken(context: Context): Promise<IToken> {
    const token = await this.getOrLoadTokenByRefreshToken(context)

    if ((await this.getClientId(context)) !== token.client.id)
      throw new InvalidClient('Client identifier does not match', context)

    const { client, user, scopes } = token
    const [
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    ] = await Promise.all([
      this.generateAccessToken(context),
      this.getAccessTokenExpiresAt(context),
      this.generateRefreshToken(context),
      this.getRefreshTokenExpiresAt(context),
    ])

    return {
      client,
      user,
      scopes,
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    }
  }

  protected async saveToken(token: IToken, context: Context): Promise<IToken> {
    return await this.saveTokenByRefreshedToken(
      token,
      await this._revokeToken(context),
      context,
    )
  }

  private async _revokeToken(context: Context) {
    const tokenToRevoke = await this.getOrLoadTokenByRefreshToken(context)

    if (!(await this.revokeToken(tokenToRevoke, context)))
      throw new InvalidGrant('The token is invalid', context)

    return tokenToRevoke
  }

  protected abstract override getTokenByRefreshToken(
    refreshToken: string,
    context: Context,
  ): PromiseToo<Maybe<IToken>>

  protected abstract saveTokenByRefreshedToken(
    token: IToken,
    refreshedToken: IToken,
    context: Context,
  ): PromiseToo<IToken>

  protected abstract revokeToken(
    token: IToken,
    context: Context,
  ): PromiseToo<boolean>
}
