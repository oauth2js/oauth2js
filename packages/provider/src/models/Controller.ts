import { is } from 'type-is'

import { AbstractExchanger } from '../abstracts/AbstractExchanger'
import { AbstractFlow } from '../abstracts/AbstractFlow'
import { AbstractGrant } from '../abstracts/AbstractGrant'
import { AbstractTokenType } from '../abstracts/AbstractTokenType'
import { isOAuth2Error } from '../abstracts/OAuth2Error'
import { InvalidRequest } from '../errors/InvalidRequest'
import { ServerError } from '../errors/ServerError'
import { UnsupportedGrantType } from '../errors/UnsupportedGrantType'
import { GRANT_TYPE_REGEXP, RESPONSE_TYPE_REGEXP } from '../helpers/regexps'
import {
  ArrayInfer,
  ArrayReadOnlyToo,
  ConstructorOf,
  Maybe,
} from '../helpers/types'
import { IMPLEMENTATION_TYPE } from '../interfaces/IControllerImplementation'
import { IRequest } from '../interfaces/IRequest'
import { IResponse } from '../interfaces/IResponse'
import { ITicket } from '../interfaces/ITicket'
import { IUser } from '../interfaces/IUser'
import { TokenBearerType } from '../standards/TokenBearerType'
import { Context } from './Context'

export type Index = [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  -1,
]

export type WalkAnd<
  T extends ArrayReadOnlyToo<ConstructorOf<unknown>>,
  K extends string | symbol,
  I extends number = 0,
> = I extends -1
  ? never
  : T[I] extends undefined
  ? never
  : K extends keyof InstanceType<T[I]>
  ? T[Index[I]] extends undefined
    ? InstanceType<T[I]>[K]
    : InstanceType<T[I]>[K] | WalkAnd<T, K, Index[I]>
  : never

export type ImplementationOf<
  G extends
    | ArrayReadOnlyToo<ConstructorOf<AbstractGrant>>
    | ConstructorOf<AbstractGrant>,
> = G extends ArrayReadOnlyToo<ConstructorOf<AbstractGrant>>
  ? Partial<WalkAnd<G, IMPLEMENTATION_TYPE>>
  : G extends ConstructorOf<AbstractGrant>
  ? Partial<WalkAnd<[G], IMPLEMENTATION_TYPE>>
  : never

export class Controller<
  G extends ArrayReadOnlyToo<ConstructorOf<AbstractGrant>> = [],
> {
  private _grants: AbstractGrant[] = []
  private _flows: AbstractFlow[] = []
  private _tokenTypes: AbstractTokenType[] = [new TokenBearerType()]

  public use<T extends ConstructorOf<AbstractGrant>>(
    Grant: T,
  ): Controller<[...G, T]> {
    const grant = new Grant()

    this._grants.push(grant)

    grant['whenBeingAttached'](this as Controller)

    if (grant instanceof AbstractExchanger && 'flow' in grant) {
      const flow = grant['flow']

      this._flows.push(flow)

      flow['whenBeingAttached'](this as Controller)
    }

    return this as unknown as Controller<[...G, T]>
  }

  public implement(implementations: ImplementationOf<G>): this
  public implement<T extends ArrayReadOnlyToo<ArrayInfer<G>>>(
    implementations: ImplementationOf<T>,
    targets: T,
  ): this
  public implement(
    implementations: ImplementationOf<G>,
    targets?: Array<ConstructorOf<AbstractGrant>>,
  ): this {
    if (targets)
      this._grants
        .filter(grant =>
          targets.includes(grant.constructor as ConstructorOf<AbstractGrant>),
        )
        .forEach(instance => Object.assign(instance, implementations))
    else
      this._grants.forEach(instance => Object.assign(instance, implementations))

    return this
  }

  public async token(request: IRequest, response: IResponse): Promise<void> {
    const context = new Context(request, response)

    try {
      if (
        !(
          is(
            context.request.headers['content-type'] as string,
            'application/x-www-form-urlencoded',
          ) ||
          is(
            context.request.headers['Content-Type'] as string,
            'application/x-www-form-urlencoded',
          )
        )
      )
        throw new InvalidRequest(
          "Requests for an access token must be in the form 'application/x-www-form-urlencoded'",
          context,
        )

      const grantType = request.body.grant_type

      if (!grantType)
        throw new InvalidRequest('The grant type is missing', context)

      if (typeof grantType !== 'string')
        throw new InvalidRequest(
          `The grant type must be of type string but is set to ${typeof grantType}`,
          context,
        )

      // TODO: Invalid request or grant?
      if (!GRANT_TYPE_REGEXP.test(grantType))
        throw new InvalidRequest('The grant type is invalid', context)

      const grant = this._grants.find(grant => grant['type'] === grantType)

      if (!grant)
        throw new UnsupportedGrantType(
          'The grant type is not currently supported by the server.',
          context,
        )

      const token = await AbstractGrant.token.call(grant, context)

      const tokenType = this._tokenTypes.find(
        tokenType => tokenType.type === 'bearer',
      )

      if (!tokenType)
        throw new ServerError('No response type implemented on server', context)

      context.response.setHeader('Content-Type', 'application/json')
      context.response.status(200)
      context.response.send(JSON.stringify(tokenType.toJSON(token)))
    } catch (error: unknown) {
      if (isOAuth2Error(error)) {
        context.response.setHeader('Content-Type', 'application/json')
        context.response.status(error.status)
        context.response.send(
          JSON.stringify({
            error: error.error,
            error_description: error.description,
          }),
        )
      } else throw error
    }
  }

  public async release(request: IRequest, response: IResponse): Promise<void> {
    const context = new Context(request, response)
    const responseType =
      request.body.response_type ?? request.query.response_type

    if (!responseType)
      throw new InvalidRequest('The response type is missing', context)

    if (typeof responseType !== 'string')
      throw new InvalidRequest(
        `The response type must be of type string but is set to ${typeof responseType}`,
        context,
      )

    // TODO: Invalid request or grant?
    if (!RESPONSE_TYPE_REGEXP.test(responseType))
      throw new InvalidRequest('The response type is invalid', context)

    const flow = this._flows.find(flow => flow['type'] === responseType)

    if (!flow)
      throw new UnsupportedGrantType(
        'The response type is not currently supported by the server.',
        context,
      )

    await AbstractFlow.release.call(flow, context).catch(async error => {
      if (!isOAuth2Error(error)) throw error
      await AbstractFlow.releaseError.call(flow, error, context)
    })
  }

  public async authorize(
    exchanger: ConstructorOf<AbstractExchanger>,
    ticket: ITicket,
    user: IUser,
    request: IRequest,
    response: IResponse,
  ) {
    const context = new Context(request, response)

    const grant = this._grants.find(
      grant => grant instanceof exchanger,
    ) as Maybe<AbstractExchanger>

    if (!grant)
      throw new ServerError(
        `No instance of ${exchanger.name} grant found`,
        context,
      )

    const flow = grant['flow'] as Maybe<AbstractFlow>

    if (!flow)
      throw new ServerError(
        `The ${exchanger.name} grant has no ticket flow.`,
        context,
      )

    await AbstractFlow.authorize
      .call(flow, ticket, user, context)
      .catch(async error => {
        if (!isOAuth2Error(error)) throw error
        await AbstractFlow.authorizeError.call(
          flow,
          error,
          ticket,
          user,
          context,
        )
      })
  }

  public async deny(
    exchanger: ConstructorOf<AbstractExchanger>,
    ticket: ITicket,
    request: IRequest,
    response: IResponse,
  ) {
    const context = new Context(request, response)
    const grant = this._grants.find(
      grant => grant instanceof exchanger,
    ) as Maybe<AbstractExchanger>

    if (!grant)
      throw new ServerError(
        `No instance of ${exchanger.name} grant found`,
        context,
      )

    const flow = grant['flow'] as Maybe<AbstractFlow>

    if (!flow)
      throw new ServerError(
        `The ${exchanger.name} grant has no ticket flow.`,
        context,
      )

    await AbstractFlow.deny.call(flow, ticket, context).catch(async error => {
      if (!isOAuth2Error(error)) throw error
      await AbstractFlow.denyError.call(flow, error, ticket, context)
    })
  }
}
