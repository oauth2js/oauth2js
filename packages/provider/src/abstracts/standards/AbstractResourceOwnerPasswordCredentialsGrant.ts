import { Mixed } from '@prunus/mixin'

import {
  DEFAULT_ACCESS_TOKEN_LIFETIME,
  DEFAULT_REFRESH_TOKEN_LIFETIME,
} from '../../helpers/defaults'
import { Maybe, PromiseToo } from '../../helpers/types'
import { IClient } from '../../interfaces/IClient'
import { RequestLocation } from '../../interfaces/IRequest'
import { IScope } from '../../interfaces/IScope'
import { IToken } from '../../interfaces/IToken'
import { IUser } from '../../interfaces/IUser'
import { Context } from '../../models/Context'
import {
  ClientCredentials,
  UseClientByCredentials,
} from '../../traits/UseClientByCredentials'
import { UseClientById } from '../../traits/UseClientById'
import { UseClientPolymorph } from '../../traits/UseClientPolymorph'
import { UseScopes } from '../../traits/UseScopes'
import {
  UserCredentials,
  UseUserByCredentials,
} from '../../traits/UseUserByCredentials'
import { AbstractGrant } from '../AbstractGrant'

export abstract class AbstractResourceOwnerPasswordCredentialsGrant extends Mixed(
  UseClientPolymorph(UseClientByCredentials, UseClientById),
  AbstractGrant,
  UseUserByCredentials,
  UseScopes,
) {
  protected override readonly type = 'password'
  protected readonly scopeLocation = RequestLocation.Body
  protected readonly refreshTokenLifetime = DEFAULT_REFRESH_TOKEN_LIFETIME
  protected readonly accessTokenLifetime = DEFAULT_ACCESS_TOKEN_LIFETIME
  protected readonly userCredentialsLocation: RequestLocation =
    RequestLocation.Body
  protected readonly clientCredentialsLocation: RequestLocation =
    RequestLocation.Body
  protected readonly clientIdLocation: RequestLocation = RequestLocation.Body

  protected async createToken(context: Context): Promise<IToken> {
    const client = await this.getOrLoadClient(context)
    const scopes = await this.getOrLoadScopesObjects(context)
    const user = await this.loadUser(context)

    const [
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    ] = await Promise.all([
      this.generateAccessToken(context),
      this.getAccessTokenExpiresAt(context),
      this.generateRefreshToken(context),
      this.getRefreshTokenExpiresAt(context),
    ])

    return {
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
      client,
      scopes,
      user,
    }
  }

  protected abstract override selectUseClientProvider(
    context: Context,
  ): typeof UseClientByCredentials | typeof UseClientById

  protected abstract override getClientByCredentials(
    credentials: ClientCredentials,
    context: Context,
  ): PromiseToo<Maybe<IClient>>

  protected abstract override getClientById(
    clientId: string,
    context: Context,
  ): PromiseToo<Maybe<IClient>>

  protected abstract override getScopesObjects(
    scopes: string[],
    context: Context,
  ): PromiseToo<IScope[]>

  protected abstract override getUserByCredentials(
    credentials: UserCredentials,
    context: Context,
  ): PromiseToo<Maybe<IUser>>

  protected abstract override saveToken(
    token: IToken,
    context: Context,
  ): PromiseToo<IToken>
}
