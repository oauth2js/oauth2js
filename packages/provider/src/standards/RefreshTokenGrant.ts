import { AbstractRefreshTokenGrant } from '../abstracts/standards/AbstractRefreshTokenGrant'
import { ServerError } from '../errors/ServerError'
import { PromiseToo, Maybe } from '../helpers/types'
import { IControllerImplementation } from '../interfaces/IControllerImplementation'
import { IToken } from '../interfaces/IToken'
import { Context } from '../models/Context'

export class RefreshTokenGrant extends AbstractRefreshTokenGrant {
  // @ts-expect-error: The `saveToken` must remain in the constructor implementation but for controller implementation issues it will be redirected to `saveExchangeToken`.
  protected override get saveToken() {
    return super.saveToken
  }
  protected override set saveToken(saveToken) {
    this.saveTokenByRefreshedToken = async function (
      token,
      refreshedToken,
      context,
    ) {
      return saveToken(token, context)
    }
  }

  protected getTokenByRefreshToken(
    refreshToken: string,
    context: Context,
  ): PromiseToo<Maybe<IToken>> {
    throw new ServerError(
      'The `getTokenByRefreshToken` definition is missing from the implementation.',
      context,
    )
  }

  protected saveTokenByRefreshedToken(
    token: IToken,
    refreshedToken: IToken,
    context: Context,
  ): PromiseToo<IToken> {
    throw new ServerError(
      'The `saveTokenByRefreshedToken` definition is missing from the implementation.',
      context,
    )
  }

  protected revokeToken(token: IToken, context: Context): PromiseToo<boolean> {
    throw new ServerError(
      'The `revokeToken` definition is missing from the implementation.',
      context,
    )
  }
}

export interface RefreshTokenGrant
  extends IControllerImplementation<{
    getTokenByRefreshToken(
      refreshToken: string,
      context: Context,
    ): PromiseToo<Maybe<IToken>>
    saveTokenByRefreshedToken(
      token: IToken,
      refreshedToken: IToken,
      context: Context,
    ): PromiseToo<IToken>
    revokeToken(token: IToken, context: Context): PromiseToo<boolean>
    saveToken(token: IToken, context: Context): Promise<IToken>
  }> {}
