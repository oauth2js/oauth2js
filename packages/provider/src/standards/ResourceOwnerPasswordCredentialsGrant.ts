import { AbstractResourceOwnerPasswordCredentialsGrant } from '../abstracts/standards/AbstractResourceOwnerPasswordCredentialsGrant'
import { ServerError } from '../errors/ServerError'
import { PromiseToo, Maybe } from '../helpers/types'
import { IClient } from '../interfaces/IClient'
import { IControllerImplementation } from '../interfaces/IControllerImplementation'
import { IScope } from '../interfaces/IScope'
import { IToken } from '../interfaces/IToken'
import { IUser } from '../interfaces/IUser'
import { Context } from '../models/Context'
import {
  ClientCredentials,
  UseClientByCredentials,
} from '../traits/UseClientByCredentials'
import { UseClientById } from '../traits/UseClientById'
import { UserCredentials } from '../traits/UseUserByCredentials'

export class ResourceOwnerPasswordCredentialsGrant extends AbstractResourceOwnerPasswordCredentialsGrant {
  protected selectUseClientProvider(
    context: Context,
  ): typeof UseClientByCredentials | typeof UseClientById {
    if (
      this.getClientByCredentials !==
      ResourceOwnerPasswordCredentialsGrant.prototype.getClientByCredentials
    )
      if (
        this.getClientById !==
        ResourceOwnerPasswordCredentialsGrant.prototype.getClientById
      )
        throw new ServerError(
          'Unable to resolve which client provider to use, conflict when `getClientByCredentials` and `getClientById` are set without a custom `selectUseClientProvider`.',
          context,
        )
      else return UseClientByCredentials

    if (
      this.getClientById !==
      ResourceOwnerPasswordCredentialsGrant.prototype.getClientById
    )
      return UseClientById

    throw new ServerError(
      'Unable to resolve which client provider to use, conflict when neither `getClientByCredentials` nor `getClientById` are defined.',
      context,
    )
  }

  protected getClientByCredentials(
    credentials: ClientCredentials,
    context: Context,
  ): PromiseToo<Maybe<IClient>> {
    throw new ServerError(
      'The `getClientByCredentials` definition is missing from the implementation.',
      context,
    )
  }

  protected getClientById(
    clientId: string,
    context: Context,
  ): PromiseToo<Maybe<IClient>> {
    throw new ServerError(
      'The `getClientById` definition is missing from the implementation.',
      context,
    )
  }

  protected getScopesObjects(
    scopes: string[],
    context: Context,
  ): PromiseToo<IScope[]> {
    throw new ServerError(
      'The `getScopesObjects` definition is missing from the implementation.',
      context,
    )
  }

  protected getUserByCredentials(
    credentials: UserCredentials,
    context: Context,
  ): PromiseToo<Maybe<IUser>> {
    throw new ServerError(
      'The `getUserByCredentials` definition is missing from the implementation.',
      context,
    )
  }

  protected saveToken(token: IToken, context: Context): PromiseToo<IToken> {
    throw new ServerError(
      'The `saveToken` definition is missing from the implementation.',
      context,
    )
  }
}

export interface ResourceOwnerPasswordCredentialsGrant
  extends IControllerImplementation<{
    getScopesObjects(scopes: string[], context: Context): PromiseToo<IScope[]>
    getUserByCredentials(
      credentials: UserCredentials,
      context: Context,
    ): PromiseToo<Maybe<IUser>>
    saveToken(token: IToken, context: Context): PromiseToo<IToken>
    getClientByCredentials?(
      credentials: ClientCredentials,
      context: Context,
    ): PromiseToo<Maybe<IClient>>
    getClientById?(
      clientId: string,
      context: Context,
    ): PromiseToo<Maybe<IClient>>
  }> {}
