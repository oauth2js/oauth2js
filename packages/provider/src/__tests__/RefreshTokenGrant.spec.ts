import { MockController } from '../../.jest/tools/Controller'
import { Database } from '../../.jest/tools/Database'
import {
  DatabaseClient,
  DatabaseScope,
  DatabaseUser,
} from '../../.jest/tools/DatabaseTypes'
import { IToken } from '../interfaces/IToken'
import { RefreshTokenGrant } from '../standards/RefreshTokenGrant'
import { ResourceOwnerPasswordCredentialsGrant } from '../standards/ResourceOwnerPasswordCredentialsGrant'

describe('Refresh Token Grant', () => {
  it('generate token', async () => {
    const db = Database()

    db.clients.insertOne({
      id: 'client-test-id',
      secret: 'client-test-secret',
      redirectUris: ['http://localhost:8080'],
    })
    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller =>
      controller
        .use(ResourceOwnerPasswordCredentialsGrant)
        .use(RefreshTokenGrant)
        .implement({
          getClientByCredentials: credentials =>
            db.clients.findOne<DatabaseClient>(
              client =>
                client.id === credentials.id &&
                client.secret === credentials.secret,
            ),
          getScopesObjects: scopes =>
            db.scopes.find<DatabaseScope>(scope =>
              scopes.includes(scope.alias),
            ),
          getUserByCredentials: credentials =>
            db.users.findOne<DatabaseUser>(
              user =>
                user.username === credentials.username &&
                user.password === credentials.password,
            ),
          saveToken: token => db.tokens.insertOne(token),
          getTokenByRefreshToken: refreshToken =>
            db.tokens.findOne<IToken>(
              token => token.refreshToken === refreshToken,
            ),
          revokeToken: ({ accessToken }) =>
            !!db.tokens.softDeleteOne(
              token => token.accessToken === accessToken,
            ),
          saveTokenByRefreshedToken: (token, refreshedToken) =>
            db.tokens.insertOne({ ...token, token: refreshedToken }),
        }),
    )

    const headers = {
      Authorization: `Basic ${Buffer.from(
        'client-test-id:client-test-secret',
      ).toString('base64')}`,
    }
    const body = {
      grant_type: 'password',
      username: 'user-test-username',
      password: 'user-test-password',
      scope: 'read+write',
    }
    let response = await controller.token({ headers, body })

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      access_token: expect.any(String),
      refresh_token: expect.any(String),
    })

    expect(db.tokens.count()).toBe(1)
    expect(db.tokens.findOne()).toMatchSnapshot({
      accessToken: expect.any(String),
      accessTokenExpiresAt: expect.any(Date),
      refreshToken: expect.any(String),
      refreshTokenExpiresAt: expect.any(Date),
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    })

    response = await controller.token({
      body: {
        grant_type: 'refresh_token',
        refresh_token: db.tokens.findOneOrFail().refreshToken,
        client_id: 'client-test-id',
      },
    })

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      access_token: expect.any(String),
      refresh_token: expect.any(String),
    })
    expect(db.tokens.count({ includeDeleted: true })).toBe(2)
    expect(db.tokens.findOne()).toMatchSnapshot({
      accessToken: expect.any(String),
      accessTokenExpiresAt: expect.any(Date),
      refreshToken: expect.any(String),
      refreshTokenExpiresAt: expect.any(Date),
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      token: {
        accessToken: expect.any(String),
        accessTokenExpiresAt: expect.any(Date),
        refreshToken: expect.any(String),
        refreshTokenExpiresAt: expect.any(Date),
        scopes: [
          {
            createdAt: expect.any(Date),
            updatedAt: expect.any(Date),
          },
          {
            createdAt: expect.any(Date),
            updatedAt: expect.any(Date),
          },
        ],
        client: {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        user: {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        deletedAt: expect.any(Date),
      },
    })
  })

  it('with error when getTokenByRefreshToken is not implemented', async () => {
    const db = Database()

    db.clients.insertOne({
      id: 'client-test-id',
      secret: 'client-test-secret',
      redirectUris: ['http://localhost:8080'],
    })
    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller =>
      controller
        .use(ResourceOwnerPasswordCredentialsGrant)
        .use(RefreshTokenGrant)
        .implement({
          getClientByCredentials: credentials =>
            db.clients.findOne<DatabaseClient>(
              client =>
                client.id === credentials.id &&
                client.secret === credentials.secret,
            ),
          getScopesObjects: scopes =>
            db.scopes.find<DatabaseScope>(scope =>
              scopes.includes(scope.alias),
            ),
          getUserByCredentials: credentials =>
            db.users.findOne<DatabaseUser>(
              user =>
                user.username === credentials.username &&
                user.password === credentials.password,
            ),
          saveToken: token => db.tokens.insertOne(token),
        }),
    )

    const headers = {
      Authorization: `Basic ${Buffer.from(
        'client-test-id:client-test-secret',
      ).toString('base64')}`,
    }
    const { body } = await controller.tokenOrFail({
      headers,
      body: {
        grant_type: 'password',
        username: 'user-test-username',
        password: 'user-test-password',
        scope: 'read+write',
      },
    })

    const { refresh_token: refreshToken } = JSON.parse(body)
    const response = await controller.token({
      body: {
        grant_type: 'refresh_token',
        refresh_token: refreshToken,
        client_id: 'client-test-id',
      },
    })

    expect(response.status()).toBe(500)
    expect(JSON.parse(response.body)).toMatchSnapshot()
  })

  it.todo(
    'with error when saveToken or saveTokenByRefreshedToken is not implemented',
  )

  it('with error when revokeToken is not implemented', async () => {
    const db = Database()

    db.clients.insertOne({
      id: 'client-test-id',
      secret: 'client-test-secret',
      redirectUris: ['http://localhost:8080'],
    })
    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller =>
      controller
        .use(ResourceOwnerPasswordCredentialsGrant)
        .use(RefreshTokenGrant)
        .implement({
          getClientByCredentials: credentials =>
            db.clients.findOne<DatabaseClient>(
              client =>
                client.id === credentials.id &&
                client.secret === credentials.secret,
            ),
          getScopesObjects: scopes =>
            db.scopes.find<DatabaseScope>(scope =>
              scopes.includes(scope.alias),
            ),
          getUserByCredentials: credentials =>
            db.users.findOne<DatabaseUser>(
              user =>
                user.username === credentials.username &&
                user.password === credentials.password,
            ),
          saveToken: token => db.tokens.insertOne(token),
          getTokenByRefreshToken: refreshToken =>
            db.tokens.findOne<IToken>(
              token => token.refreshToken === refreshToken,
            ),
        }),
    )

    const headers = {
      Authorization: `Basic ${Buffer.from(
        'client-test-id:client-test-secret',
      ).toString('base64')}`,
    }
    const { body } = await controller.tokenOrFail({
      headers,
      body: {
        grant_type: 'password',
        username: 'user-test-username',
        password: 'user-test-password',
        scope: 'read+write',
      },
    })

    const { refresh_token: refreshToken } = JSON.parse(body)
    const response = await controller.token({
      body: {
        grant_type: 'refresh_token',
        refresh_token: refreshToken,
        client_id: 'client-test-id',
      },
    })

    expect(response.status()).toBe(500)
    expect(JSON.parse(response.body)).toMatchSnapshot()
  })
})
