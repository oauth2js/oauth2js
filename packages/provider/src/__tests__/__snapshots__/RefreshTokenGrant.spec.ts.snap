// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Refresh Token Grant generate token 1`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Refresh Token Grant generate token 2`] = `
Object {
  "access_token": Any<String>,
  "expires_in": 7200,
  "refresh_token": Any<String>,
  "token_type": "Bearer",
}
`;

exports[`Refresh Token Grant generate token 3`] = `
Object {
  "accessToken": Any<String>,
  "accessTokenExpiresAt": Any<Date>,
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "createdAt": Any<Date>,
  "deletedAt": null,
  "refreshToken": Any<String>,
  "refreshTokenExpiresAt": Any<Date>,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "updatedAt": Any<Date>,
  "user": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<Date>,
    "username": "user-test-username",
  },
}
`;

exports[`Refresh Token Grant generate token 4`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Refresh Token Grant generate token 5`] = `
Object {
  "access_token": Any<String>,
  "expires_in": 7200,
  "refresh_token": Any<String>,
  "token_type": "Bearer",
}
`;

exports[`Refresh Token Grant generate token 6`] = `
Object {
  "accessToken": Any<String>,
  "accessTokenExpiresAt": Any<Date>,
  "client": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<Date>,
  },
  "createdAt": Any<Date>,
  "deletedAt": null,
  "refreshToken": Any<String>,
  "refreshTokenExpiresAt": Any<Date>,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<Date>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<Date>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<Date>,
    },
  ],
  "token": Object {
    "accessToken": Any<String>,
    "accessTokenExpiresAt": Any<Date>,
    "client": Object {
      "createdAt": Any<Date>,
      "deletedAt": null,
      "id": "client-test-id",
      "redirectUris": Array [
        "http://localhost:8080",
      ],
      "secret": "client-test-secret",
      "updatedAt": Any<Date>,
    },
    "createdAt": Any<Date>,
    "deletedAt": Any<Date>,
    "refreshToken": Any<String>,
    "refreshTokenExpiresAt": Any<Date>,
    "scopes": Array [
      Object {
        "alias": "read",
        "createdAt": Any<Date>,
        "deletedAt": null,
        "description": "Read",
        "updatedAt": Any<Date>,
      },
      Object {
        "alias": "write",
        "createdAt": Any<Date>,
        "deletedAt": null,
        "description": "Write",
        "updatedAt": Any<Date>,
      },
    ],
    "updatedAt": Any<Date>,
    "user": Object {
      "createdAt": Any<Date>,
      "deletedAt": null,
      "password": "user-test-password",
      "updatedAt": Any<Date>,
      "username": "user-test-username",
    },
  },
  "updatedAt": Any<Date>,
  "user": Object {
    "createdAt": Any<Date>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<Date>,
    "username": "user-test-username",
  },
}
`;

exports[`Refresh Token Grant with error when getTokenByRefreshToken is not implemented 1`] = `
Object {
  "error": "server_error",
  "error_description": "The \`getTokenByRefreshToken\` definition is missing from the implementation.",
}
`;

exports[`Refresh Token Grant with error when revokeToken is not implemented 1`] = `
Object {
  "error": "server_error",
  "error_description": "The \`revokeToken\` definition is missing from the implementation.",
}
`;
